src/dm/DispatchManagerActions.cc

```diff
@@ -2249,7 +2442,750 @@ int DispatchManager::live_updateconf(std::unique_ptr<VirtualMachine> vm,
+
+        switch (lstate)
+        {
+            //Cannnot update VM, SG rules being updated/created
+            case VirtualMachine::BOOT:
+            case VirtualMachine::BOOT_MIGRATE:
+            case VirtualMachine::BOOT_SUSPENDED:
+            case VirtualMachine::BOOT_STOPPED:
+            case VirtualMachine::BOOT_UNDEPLOY:
+            case VirtualMachine::BOOT_POWEROFF:
+            case VirtualMachine::BOOT_UNKNOWN:
+            case VirtualMachine::BOOT_FAILURE:
+            case VirtualMachine::BOOT_MIGRATE_FAILURE:
+            case VirtualMachine::BOOT_UNDEPLOY_FAILURE:
+            case VirtualMachine::BOOT_STOPPED_FAILURE:
+            case VirtualMachine::MIGRATE:
+            case VirtualMachine::HOTPLUG_NIC:
+            case VirtualMachine::HOTPLUG_NIC_POWEROFF:
+            case VirtualMachine::UNKNOWN:
+                oss << "VM " << vid << " is in wrong state " << vm->state_str();
+                error_str = oss.str();
+
+                return -1;
+
+            //Update SG rules at host
+            case VirtualMachine::RUNNING:
+            case VirtualMachine::HOTPLUG:
+            case VirtualMachine::HOTPLUG_SNAPSHOT:
+            case VirtualMachine::HOTPLUG_SAVEAS:
+            case VirtualMachine::HOTPLUG_RESIZE:
+            case VirtualMachine::DISK_SNAPSHOT:
+            case VirtualMachine::DISK_SNAPSHOT_DELETE:
+            case VirtualMachine::DISK_RESIZE:
+                is_update = true;
+                break;
+
+            default:
+                break;
+        }
+
+        auto nic = vm->get_nic(nicid);
+
+        if ( nic == nullptr )
+        {
+            oss << "VM " << vid << " doesn't have NIC id " << nicid;
+            error_str = oss.str();
+
+            return -1;
+        }
+
+        set<int> sgs;
+        nic->get_security_groups(sgs);
+
+        if (sgs.find(sgid) != sgs.end())
+        {
+            oss << "VM " << vid << " SG " << sgid << " already in NIC " << nicid;
+            error_str = oss.str();
+
+            return -1;
+        }
+
+        // ----------------------------------------------------------------------
+        // Update VM data
+        // - NIC to include new SG
+        // - Template to include new SG rules (note can be there already)
+        // ----------------------------------------------------------------------
+        vector<VectorAttribute *> sg_rules;
+
+        sgpool->get_security_group_rules(-1, sgid, sg_rules);
+
+        nic->add_security_group(sgid);
+
+        vm->remove_security_group(sgid); //duplicates
+
+        vm->add_template_attribute(sg_rules);
+
+        vmpool->update(vm.get());
+
+        if ( is_update )
+        {
+            vmm->updatesg(vm.get(), sgid);
+        }
+    }
+    else
+    {
+        oss << "Could not attach SG to VM " << vid << ", VM does not exist";
+        error_str = oss.str();
+
+        return -1;
+    }
+
+    if (auto sg = sgpool->get(sgid))
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

We observed that you are using unintended operations or assignments that could lead to logical errors, vulnerabilities, or undefined behavior. To prevent these issues, ensure that you follow best practices, such as using explicit comparisons and clear operator usage, and using parentheses to clarify intent. Learn more - https://cvefeed.io/cwe/detail/cwe-480-use-of-incorrect-operator